node/src/components/block_synchronizer/tests.rs

```diff
@@ -336,3 +452,1231 @@ async fn should_not_stall_after_registering_new_era_validator_weights() {
+
+    // Register block for fwd sync
+    assert!(block_synchronizer.register_block_by_hash(*block.hash(), false, true));
+    assert!(block_synchronizer.forward.is_some());
+    block_synchronizer.register_peers(*block.hash(), peers.clone());
+    let fwd_builder = block_synchronizer
+        .forward
+        .as_mut()
+        .expect("Forward builder should have been initialized");
+    assert!(fwd_builder
+        .register_block_header(block.clone().take_header(), None)
+        .is_ok());
+    fwd_builder.register_era_validator_weights(&block_synchronizer.validator_matrix);
+
+    // Check the block acquisition state
+    assert_matches!(
+        fwd_builder.block_acquisition_state(),
+        BlockAcquisitionState::HaveBlockHeader(header, _) if header.block_hash() == *block.hash()
+    );
+
+    // Synchronizer should fetch finality signatures
+    let events = need_next(
+        &mut rng,
+        &mock_reactor,
+        &mut block_synchronizer,
+        std::cmp::min(num_validators, MAX_SIMULTANEOUS_PEERS), /* We have num_validators
+                                                                * validators so we
+                                                                * require the num_validators
+                                                                * signatures */
+    )
+    .await;
+
+    // Check what signatures were requested
+    let mut sigs_requested = Vec::new();
+    for event in events {
+        assert_matches!(
+            event,
+            MockReactorEvent::FinalitySignatureFetcherRequest(FetcherRequest {
+                id,
+                peer,
+                ..
+            }) => {
+                assert!(peers.contains(&peer));
+                assert_eq!(id.block_hash, *block.hash());
+                assert_eq!(id.era_id, block.header().era_id());
+                sigs_requested.push((peer, id.public_key));
+            }
+        );
+    }
+
+    // Simulate failed fetch of finality signatures
+    let mut generated_effects = Effects::new();
+    for (peer, public_key) in sigs_requested {
+        let effects = block_synchronizer.handle_event(
+            mock_reactor.effect_builder(),
+            &mut rng,
+            Event::FinalitySignatureFetched(Err(FetcherError::Absent {
+                id: Box::new(Box::new(FinalitySignatureId {
+                    block_hash: *block.hash(),
+                    era_id: block.header().era_id(),
+                    public_key,
+                })),
+                peer,
+            })),
+        );
+        // the effects array should be empty while the latch is active
+        // once the latch is reset, we should get some effects
+        generated_effects.extend(effects);
+    }
+
+    // If the effects are empty at this point, then the synchronizer gets stuck
+    assert!(!generated_effects.is_empty());
+}
+
+#[tokio::test]
+async fn next_action_for_have_weak_finality_is_fetching_block_body() {
```

Comment:
Judging by just the name, this is very similar to [this guy](https://github.com/casper-network/casper-node/pull/3816/files#diff-ecca1f5c3a37ff0c7cb64100807c2e3074f1dfdd68582dff2dd768f7a313806bR726).

The latter one is more sophisticated, but please check if we don't duplicate some parts of the test unnecessarily.

Reply:
You're right, these are similar but actually their target is different. This one puts the builder in `HaveWeakFinality` and checks if the next action is fetching the block body. The other test wants to check that fetching the block body only happens after weak finality was reached, and it goes through some steps of registering signatures to see if it doesn't transition too early. I'll rename the test that you referenced to better state its purpose.